### PR TITLE
Implement url joining utility function

### LINKED ISSUE #955: Wanted: uri joining utility function in rust
To work towards #866, we need to have a utility function which can do URL path resolution (resolving ".." and "." properly)

We currently use [mozilla's url crate](https://github.com/servo/rust-url) for this, but in order to avoid having two different URL parsing crates, I would like to use Hyper's [http::uri crate](https://docs.rs/http/0.1.13/http/uri/index.html) instead. 

However, this crate has less functionality - in particular we are missing a "join" function. 

I've outlined a first attempt at this below... but it doesn't handle ".." and "." resolution. It needs tests.

```rust
fn uri_join(
  base: uri::Uri,
  path: &str,
) -> Result<uri::Uri, http::Error> {
  let base_parts = base.into_parts();
  let base_path_and_query = base_parts.path_and_query.unwrap();
  let base_path = base_path_and_query.path();
  let joined_path = String::from(base_path) + "/" + path;
  let joined_path_and_query =
    uri::PathAndQuery::from_shared(joined_path.into()).unwrap();
  let mut joined_parts = uri::Parts::default();
  joined_parts.scheme = base_parts.scheme;
  joined_parts.authority = base_parts.authority;
  joined_parts.path_and_query = Some(joined_path_and_query);

  uri::Uri::from_parts(joined_parts)
}
```

This could even be released as an independent crate - maybe using the name "uri_util" - sent upstream to [the http crate](https://github.com/hyperium/http) - or put into deno in src/uri_util.rs.

### PR DESCRIPTION
**Pull request by romandev**
Fixes #955

### CODE REVIEW

File: src/uri_util.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    .parse::<uri::Uri>()
+    .unwrap();
+  assert_eq!(expected, _uri_join(base, "./hello/./world").unwrap());
+}
+
+#[test]
+fn test_uri_join_with_dotdot() {
+  let base = "https://example.com".parse::<uri::Uri>().unwrap();
+  let expected = "https://example.com/world".parse::<uri::Uri>().unwrap();
+  assert_eq!(expected, _uri_join(base, "hello/../world").unwrap());
```
Comment (ry) [inline src/uri_util.rs]: Can we get a more complex example with more than one ".." ?
Also a test where there are multiple slashes "hello///world"
And also a test where the base URL ends in a slash. EG "https://example.com/foo/bar/"

Comment (romandev) [reply src/uri_util.rs]: Sure, I'll do it soon. Thank you for review!

**Comment by ry** [comment]
Nice work! But it needs more tests. Please have a look at 
RFC 3986 Section 5.2
And these tests from Go: https://github.com/golang/go/blob/63e964e1741277c4da004e366111243f9ec942a2/src/net/url/url_test.go#L1020-L1142

**Comment by romandev** [comment]
I updated more tests. Please take another look.

**Comment by ry** [comment]
Great work!
It would be super cool if we also could resolve alternate URLs - so 

```rust
uri_join(base_uri, "http://some.com/other/host") // results in "http://some.com/other/host"
```

Maybe uri_join should return a URI instead of a string?

**Comment by romandev** [comment]
> Great work!
> It would be super cool if we also could resolve alternate URLs - so
> ```rust
> uri_join(base_uri, "http://some.com/other/host") // results in "http://some.com/other/host"
> ```

Done.

> Maybe uri_join should return a URI instead of a string?

The `uri_join()` is already return a URI instead of a string.
If you meant reference uri (second parameter), it's not possible for now.
It can be treated as InvalidUri for some relative paths.
```rust
"../../".parse::<uri::Uri>() // InvalidUri error
```

Thank you.

**Comment by ry** [comment]
> The uri_join() is already return a URI instead of a string.

Oh - sorry! I missed that.
Thank you for this critical mechanism. I will attempt to use this to remove the url crate now.

**Comment by ry** [comment]
LGTM - I will land this in #1039 - thank you!

**Comment by ctaggart** [comment]
Does `uri_join` exist anywhere anymore? I want to use this. May be I can just copy:
https://github.com/ry/deno/blob/9dfc5ecc8d222dd815fc946e83d0962f0844c538/src/uri_util.rs

**Comment by nayeemrmn** [comment]
@ctaggart Use `new URL()` (https://developer.mozilla.org/en-US/docs/Web/API/URL/URL).

**Pull request closed without merging**